fuzzers/main.py

```diff
@@ -0,0 +1,642 @@
+        """Start the fuzzer runner on corpus `Corpus`"""
+        print(f'Starting fuzzer for branch {self.branch} and target {self.target}, log is at {self.log_path}')
+
+        # Prepare the fuzz time metric
+        self.last_time = time.monotonic()
+
+        # Spin up the fuzzer process itself
+        self.proc = subprocess.Popen(
+            [
+                'cargo',
+                'fuzz',
+                'run',
+                self.target['runner'],
+                '--',
+                corpus.corpus_for(self.target),
+                corpus.artifacts_for(self.target),
+                f"-artifact_prefix={corpus.artifacts_for(self.target)}/",
+            ] + self.target['flags'],
+            cwd = self.repo_dir / self.target['crate'],
+            start_new_session = True,
+            stdout = self.log_file,
+            stderr = subprocess.STDOUT,
+        )
+
+    def poll(self):
+        """Checks if the current process is still running. Returns True if it stopped"""
+        if self.proc.poll() is None:
+            new_time = time.monotonic()
+            self.fuzz_time_metric.inc(new_time - self.last_time)
+            self.last_time = new_time
+            return False
+        else: # Fuzz crash found
+            print(f"Fuzzer running {self.target} has stopped")
+            self.fuzz_crashes_metric.inc()
+            return True
+
+    def report_crash(self, corpus: Corpus, bucket: gcs.Bucket):
+        with open(self.log_filepath, 'r') as log_file_r:
+            log_lines = log_file_r.readlines()
+
+        artifact_pattern = f'Test unit written to {corpus.artifacts_for(self.target)}/'
+        artifact = "<failed detecting relevant artifact in log file>"
+        for l in log_lines[::-1]:
+            if artifact_pattern in l:
+                artifact = l.split(artifact_pattern)[1][:-1] # Second part of the line except the \n
+                break
+
+        branch = self.branch['name']
+        logs_url = f"https://storage.cloud.google.com/fuzzer/logs/{self.log_path}"
+        downloader = f"gsutil cp gs://{bucket.name}/{self.corpus_vers}/{self.target['crate']}/{self.target['runner']}/artifacts/{artifact} {self.target['crate']}/artifacts/{self.target['runner']}/{artifact}"
+        reproducer = f"cd {self.target['crate']}\nRUSTC_BOOTSTRAP=1 cargo fuzz run {self.target['runner']} artifacts/{self.target['runner']}/{artifact}"
+        minimizer = f"cd {self.target['crate']}\nRUSTC_BOOTSTRAP=1 cargo fuzz tmin {self.target['runner']} artifacts/{self.target['runner']}/{artifact}"
+
+        # Check the artifact was not reported yet for this branch
+        global REPORTED_ARTIFACTS
+        if artifact in REPORTED_ARTIFACTS[branch]:
+            return
+        already_reported_for = []
+        for other_branch in REPORTED_ARTIFACTS.keys():
+            if artifact in REPORTED_ARTIFACTS[other_branch]:
+                already_reported_for.append(other_branch)
+        REPORTED_ARTIFACTS[branch].append(artifact)
```

Comment:
This feels like it needs a lock.

Also:

```python
        already_reported_for = [
            other_branch
            for other_branch, arts in REPORTED_ARTIFACTS.items()
            if artifact in arts
        ]
```

PS. `.keys()` is pretty much never needed.

Reply:
It would need a lock if it were multi-thread, but with the recent change there's a single thread ever doing anything (besides the inotify threads) so it shouldn't require a lock I think.

Switched to the list comprehension anyway :)